variant jumping troubleshooting (#224)
Amplitude-Developer-Docs/amplitude-dev-center: docs/experiment/guides/troubleshooting/variant-jumping.md

In Amplitude, the user ID and device ID properties are essential to identify your user and [resolve their Amplitude ID](https://help.amplitude.com/hc/en-us/articles/115003135607-Track-unique-users-in-Amplitude#h_e9913ce1-549a-4d88-a832-1f56aede581a). If the device ID or user ID used to fetch/evaluate assignments is different from the device ID and user ID used to track the exposure event, you may see variant jumping, SRMs and inconsistent or unexpected bucketing behavior.

For example, you may be sending events through a proxy or CDP which obfuscates or canonicalizes IDs before sending to Amplitude. In this case, the identity used to fetch variants would be different from the identity included in the exposure events.

Another common case is simple overlooked implementation error. For example, we've seen the following cases cause variant jumping:

* Additional characters in the ID. For example Extra quotes around the actual identity.
    * `15a4f7e9-db4e-4c57-82c7-e57a2995803a`
    * `"15a4f7e9-db4e-4c57-82c7-e57a2995803a"`
* Inconsistent capitalization, especially with UUIDs.
    * `15a4f7e9-db4e-4c57-82c7-e57a2995803a`
    * `15A4F7E9-DB4E-4C57-82C7-E57A2995803A`

> 

Reviewer: Maybe we don't have enough experience with this and we can add this later if we find anything interesting. But are there different debugging strategies for variaint jumping if people bucket by org instead of users? It seems like maybe with org there would be much less variaint jumping unless someone belongs to multiple orgs like the Unity use case

---
title: Variant Jumping
description: Learn to troubleshoot variant jumping; description, causes, and solutions.
---

Variant jumping describes when a user is exposed two or more variants for a single flag or experiment. Variant jumping may occur normally or abnormally, for various reasons. Variant jumping above a certain threshold may be cause for concern in being able to trust an analysis.

## Debugging

The best way to debug variant jumping is by identifying a user that has jumped variants, and analyzing their user timeline. If you're using remote evaluation, you should also check the Assignment event to identify assignment vs exposure discrepancies.

!!!info "To find a user who has jumped variants:"

    1. Go to the monitor tab of your experiment
    2. Find the variant jumping charts, and click "Open in Analytics"
    3. Select the bar for users who jumped variants and click "View User Streams" (you may need to use uniques instead of a formula)

    You may find it helpful to "only show" assignment and exposure events in the user timeline.

Important things to note when debugging a user timeline include:

> 1. Did you introduce any [targeting changes](#targeting-changes) while your flag/experiment was active? Could the timing of this change have effected the variant assigned to this user?

Reviewer: Maybe add that "You can see what changes were made to the flag/experiment by clicking on the "Version X" button to see the version history"

5. Updating mutual exclusion

!!!tip "Avoid variant jumping by enabling sticky bucketing"
    Enabling sticky bucketing before making targeting changes will avoid variant jumping. However, sticky bucketing may cause a [sample ratio mismatch (SRM)](./sample-ratio-mismatch.md).

### Anonymous identity merging

Variant jumping caused by anonymous identity merging may occur due to bucketing by Amplitude ID, when user profiles with distinct device IDs (resolved to different amplitude IDs) are merged when resolved to a common user ID. This may happen if a user uses your app on different devices, or if, on logout, the device ID is re-generated.

!!!info "[Learn more about Amplitude's identity resolution and merging users.](https://help.amplitude.com/hc/en-us/articles/115003135607-Track-unique-users-in-Amplitude#h_e9913ce1-549a-4d88-a832-1f56aede581a)"

 Identifying this type of variant jumping is easy by identifying the the assignment event where the user jumped between variants, and comparing the value of the Amplitude ID for both events. If the Amplitude ID is different on the two events, then there then it is very likely that anonymous identity merging was the cause.

>  To combat this type of variant jumping, consider bucketing by user ID. Note that bucking by user ID may not be possible if you expect that users in the experiment will be anonymous.

Reviewer: If they have anonymou users, is our suggestion to bucket by device id

Author: Hmmm, actually targeting by device ID may very likely increase variant jumping. Since when taking unique users in amplitude we consider users by amplitude ID, not device ID.

Author: I guess if you're certain the the user is anonymous then its ok

## Abnormal variant jumping

> Abnormal variant jumping is unexpected variant jumping that can't be explained by any [normal means](#normal-variant-jumping) Abnormal variant jumping can be tough to track down, however almost all abnormal variant jumping is caused by some for of identity mismatch: **when the user identity used in assignment is different from the identity used to track the exposure.** In short, abnormal variant jumping is almost always due to an inconsistency in implementation.

Reviewer: `normal means` -> `normal means.` (missing period)

In Amplitude, the user ID and device ID properties are essential to identify your user and [resolve their Amplitude ID](https://help.amplitude.com/hc/en-us/articles/115003135607-Track-unique-users-in-Amplitude#h_e9913ce1-549a-4d88-a832-1f56aede581a). If the device ID or user ID used to fetch/evaluate assignments is different from the device ID and user ID used to track the exposure event, you may see variant jumping, SRMs and inconsistent or unexpected bucketing behavior.

For example, you may be sending events through a proxy or CDP which obfuscates or canonicalizes IDs before sending to Amplitude. In this case, the identity used to fetch variants would be different from the identity included in the exposure events.

Another common case is simple overlooked implementation error. For example, we've seen the following cases cause variant jumping:

* Additional characters in the ID. For example Extra quotes around the actual identity.
    - `15a4f7e9-db4e-4c57-82c7-e57a2995803a`
    - `"15a4f7e9-db4e-4c57-82c7-e57a2995803a"`
* Inconsistent capitalization, especially with UUIDs.
    - `15a4f7e9-db4e-4c57-82c7-e57a2995803a`
    - `15A4F7E9-DB4E-4C57-82C7-E57A2995803A`

> 

Reviewer: Do we want to explain how to make a cohort of the users who variaint jumped and exclude then from experiment analysis? This was a possible feature we were going to add in the future.

Author: That's very interesting, not sure how to even do that!

Reviewer: I can write up the instructions tomorrow and add it to this PR

### Anonymous identity merging

Variant jumping caused by anonymous identity merging may occur due to bucketing by Amplitude ID, when user profiles with distinct device IDs (resolved to different amplitude IDs) are merged when resolved to a common user ID. This may happen if a user uses your app on different devices, or if, on logout, the device ID is re-generated.

!!!info "[Learn more about Amplitude's identity resolution and merging users.](https://help.amplitude.com/hc/en-us/articles/115003135607-Track-unique-users-in-Amplitude#h_e9913ce1-549a-4d88-a832-1f56aede581a)"

 Identifying this type of variant jumping is easy by identifying the the assignment event where the user jumped between variants, and comparing the value of the Amplitude ID for both events. If the Amplitude ID is different on the two events, then there then it is very likely that anonymous identity merging was the cause.

To combat this type of variant jumping, consider bucketing by:

* User ID: If you're only targeting user who are logged in, and have a user ID.

> * Device ID: If you're only targeting anonymous users (e.g. sign up experiment).

Reviewer 2: ```suggestion
* Device ID: If you're only targeting anonymous users. For example, in s sign up experiment.
```

## Abnormal variant jumping

Abnormal variant jumping is unexpected variant jumping that can't be explained by any [normal means](#normal-variant-jumping). Abnormal variant jumping can be tough to track down, however almost all abnormal variant jumping is caused by some for of identity mismatch: **when the user identity used in assignment is different from the identity used to track the exposure.** In short, abnormal variant jumping is almost always due to an inconsistency in implementation.

> The following examples are by no means exhaustive, but should hopefully get you thinking about the use of identity in your system with respect to Amplitude Experiment.

Reviewer 2: ```suggestion
The following examples aren't exhaustive, but should get you thinking about the use of identity in your system with respect to Amplitude Experiment.
```

### Multiple logged in accounts on a single device

> Consider this timeline for a user with multiple accounts for your app on a single device.

Reviewer 2: Please define "U1" and "U2" in the intro. Something like, 

_Consider this timeline for a user with multiple accounts (U1 and U2)..._

Reviewer 2: I realize this sounds silly, but it took me a few reads to realize it referred to the user, and we should try to reduce the chances of confusion where we can.

* Inconsistent capitalization, especially with UUIDs.
    * `15a4f7e9-db4e-4c57-82c7-e57a2995803a`
    * `15A4F7E9-DB4E-4C57-82C7-E57A2995803A`

## Removing users who variant jumped

Generally, when doing analysis you want to be careful when removing data as you may bias your results. It is better to understand why users have variant jumped and if you can fix any implementation bugs, so this does not happen again in any subsequent experiments. If you feel you have a legitimate reason for removing users who variant jumped you can follow these steps:

> 1. Go to the Monitor tab and click "Open in Analytics"

Author: Our variant jumping charts are for specific variants, wouldn't it be better to create a chart of all users who have jumped for all variants?

Reviewer: How would you do that?

2. Make a cohort of the users who variant jumped by selecting the bar for users who jumped variants and click "Create Cohort"
3. Click "Save Cohort"
4. Go to the Analyze tab and click "Open in Analytics"

> 5. In the second sister, add a where cohort != `<cohort name>` for all variants

Author: I think the "three sisters" terminology is specific to amplitude.
